src/az_ulib_ipc/az_ulib_ipc_query.c

```diff
@@ -0,0 +1,192 @@
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_writer_append_string(&jw, interface_name));
+      remaining = az_span_slice_to_end(remaining, split_pos + 1);
+      split_pos = az_span_find(remaining, AZ_SPAN_FROM_STR(","));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+    if (az_span_size(remaining) > 0)
+    {
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_writer_append_string(&jw, remaining));
+    }
+
+    AZ_ULIB_THROW_IF_AZ_ERROR(az_json_writer_append_end_array(&jw));
+
+    AZ_ULIB_THROW_IF_AZ_ERROR(az_json_writer_append_property_name(
+        &jw, AZ_SPAN_FROM_STR(QUERY_1_QUERY_CONTINUATION_TOKEN_NAME)));
+    AZ_ULIB_THROW_IF_AZ_ERROR(
+        az_json_writer_append_int32(&jw, (int32_t)model_out->continuation_token));
+    AZ_ULIB_THROW_IF_AZ_ERROR(az_json_writer_append_end_object(&jw));
+    *model_out_span = az_json_writer_get_bytes_used_in_destination(&jw);
+  }
+  AZ_ULIB_CATCH(...) { return AZ_ULIB_TRY_RESULT; }
+
+  return AZ_OK;
+}
+
+static az_result query_1_query_span_wrapper(az_span model_in_span, az_span* model_out_span)
+{
+  AZ_ULIB_TRY
+  {
+    // Unmarshalling JSON in model_in_span to query_model_in.
+    az_json_reader jr;
+    query_1_query_model_in query_model_in = { 0 };
+    AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_init(&jr, model_in_span, NULL));
+    AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    while (jr.token.kind != AZ_JSON_TOKEN_END_OBJECT)
+    {
+      if (az_json_token_is_text_equal(&jr.token, AZ_SPAN_FROM_STR(QUERY_1_QUERY_QUERY_NAME)))
+      {
+        AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+        query_model_in.query
+            = az_span_create(az_span_ptr(jr.token.slice), az_span_size(jr.token.slice));
+      }
+      AZ_ULIB_THROW_IF_AZ_ERROR(az_json_reader_next_token(&jr));
+    }
+    AZ_ULIB_THROW_IF_AZ_ERROR(AZ_ULIB_TRY_RESULT);
+
+    // Create a temporary buffer to store the query_model_out.
+    char dest_buffer[200];
+    az_span dest_span = AZ_SPAN_FROM_BUFFER(dest_buffer);
+    query_1_query_model_out query_model_out = { .result = &dest_span, .continuation_token = 0 };
+
+    // Call.
+    AZ_ULIB_THROW_IF_AZ_ERROR(query_1_query_concrete(
+        (az_ulib_model_in)&query_model_in, (az_ulib_model_out)&query_model_out));
+
+    // Marshalling query_model_out to JSON in model_out_span.
+    AZ_ULIB_THROW_IF_AZ_ERROR(marshalling_model_out_to_json(&query_model_out, model_out_span));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result query_1_next_concrete(az_ulib_model_in model_in, az_ulib_model_out model_out)
+{
+  const query_1_next_model_in* const in = (const query_1_next_model_in* const)model_in;
+  query_1_next_model_out* out = (query_1_next_model_out*)model_out;
+  out->continuation_token = in->continuation_token;
+  return az_ulib_ipc_query_next(&(out->continuation_token), out->result);
+}
+
+static az_result query_1_next_span_wrapper(az_span model_in_span, az_span* model_out_span)
```

Comment:
these span_wrapper functions are serving as the parser functions you described as our string to model in/model out interpreters, correct?

Reply:
Correct, it is necessary because we are using JSON that has the model_in and model_out as strings, not binary.